hashintel/hash
# H-355: Implement inheritance in the type editor

## 🌟 What is the purpose of this PR?

Implements the UI for entity type inheritance, allowing users to set (and remove) one or more types which another entity type 'extends', and displaying the inherited properties and links in the relevant tables.

[Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH?type=design&node-id=8377-289907&mode=design&t=cDUo38X3kTkfW0Q3-4) (internal link)

## 🔍 What does this change?




- Add `inheritsFrom` as a depth argument to entity type queries in the GraphQL API and relevant queries in the frontend
- Adds inheritance management to `@hashintel/design-system` via a new 'Extends' section at the top
- Adds a hook in the type editor to get inherited properties/links
- Adds new 'inherited property/link rows' to the properties and links tables
- **(most of the diff)** Extracts various components so that they can be used for displaying information only in the inherited rows, and aren't relying on retrieving data from the form context, since inherited properties and links are not part of the editable form data
    - this is the bulk of the diff – any deleted files and big chunks of components are just moved elsewhere, and I will comment where this is so
- Updates `use-entity-type-value` in the frontend so that it also gets properties from the inheritance chain
- Adds a 'secondary quiet' button variant to the design system (this is the 'Add type +' in the extends section)

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:




### 📜 Does this require a change to the docs?




The changes in this PR:

   - we should document inheritance 


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues




- Inherited properites / links will not appear in the entity editor

## 🐾 Next steps



- Handle inheritance in the entity editor - H-356
- Add an 'extend this type' button to existing type pages to create a new type which extends another - H412

## 🛡 What tests cover this?



- None

## ❓ How to test this?



1.  Visit a type page, try adding a parent
2. Try adding a parent to the parent
3. Try releasing a new version of a parent
4. Check that inherited properties/links display okay, aren't editable

## 📹 Demo

https://github.com/hashintel/hash/assets/37743469/95fa27d6-3b3c-42c9-a7e3-148791548fea

N.B. at one point towards the end of the video two 'Created At' appears – I fixed this in https://github.com/hashintel/hash/pull/2909/commits/ee7020e38a91310449abc63fbf197ddaf1c701e2
